server/resp/src/main/java/org/infinispan/server/resp/hll/HyperLogLog.java

```diff
@@ -0,0 +1,111 @@
+import org.infinispan.server.resp.hll.internal.CompactSet;
+import org.infinispan.server.resp.hll.internal.ExplicitSet;
+import org.infinispan.server.resp.hll.internal.HLLRepresentation;
+
+/**
+ * The HyperLogLog algorithm implementation.
+ *
+ * <p>
+ * The HyperLogLog is a probabilistic algorithm for cardinality estimation by Flajolet et al. [1]. As with other
+ * probabilistic data structures, the algorithm trades space by accuracy. It trades the memory necessary to keep track
+ * of unique items by using "short bytes" to count elements.
+ * </p>
+ *
+ * <p>
+ * The version implemented here takes inspiration from the version implemented in Redis [2]. Follows part of the Google
+ * paper [3], which makes a practical implementation, and the work on [4], with optimizations to the estimation accuracy.
+ * </p>
+ *
+ * <p><h2>How it works</h2></p>
+ *
+ * From a bird's eye view, the algorithm uses a hash function that returns a value with (P + Q) bits. The P bits identify
+ * one bucket to place the count, and the Q we extract the number of consecutive trailing zeroes. Since the higher the
+ * number of zeroes, the lower the probability (N + 1 zeroes is half the N), the algorithm then collects the values from
+ * all the buckets to guess the cardinality.
+ *
+ * <p><h2>The implementation</h2></p>
+ *
+ * We use the {@link org.infinispan.commons.hash.MurmurHash3} with 64 bits. Therefore, we have <code>P=14</code> for the
+ * number of buckets and <code>Q=64-P</code> to count the number of trailing zeroes. In theory, this allows us to
+ * count up to 2<sup>64</sup> elements!
+ *
+ * <p>
+ * We use two representations to store the elements. The first one, for small cardinalities (0 - 192), uses the explicit
+ * hash return and a hash set, as seen on {@link ExplicitSet}. The compact representation with
+ * {@link CompactSet}. The form the algorithm proposes with the "short bytes" and
+ * occupies low memory.
+ * </p>
+ *
+ * <p>
+ * The compact representation occupies roughly 12Kb and could count up to 2<sup>64</sup> elements. The explicit representation
+ * counts up to 192 because it starts using more than 12Kb of memory. After reaching the threshold, the representation
+ * changes from explicit to compact.
+ * </p>
+ *
+ * @see <a href="https://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf">[1] HyperLogLog: the analysis of a near-optimal cardinality estimation algorithm</a>
+ * @see <a href="http://antirez.com/news/75">[2] Redis new data structure: the HyperLogLog</a>
+ * @see <a href="http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/40671.pdf">[3] HyperLogLog in Practice: Algorithmic Engineering of a State of The Art Cardinality Estimation Algorithm</a>
+ * @see <a href="https://arxiv.org/pdf/1702.01284.pdf">[4] New cardinality estimation algorithms for HyperLogLog sketches</a>
+ * @since 15.0
+ * @author José Bolina
+ */
+@ProtoTypeId(ProtoStreamTypeIds.RESP_HYPER_LOG_LOG)
+public class HyperLogLog {
+
```

Comment:
would this be interesting for hotrod? if the answer is yes, we should move it somewhere else, to another dependency if multimap dependency is not the right place